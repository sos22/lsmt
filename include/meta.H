#ifndef META_H__
#define META_H__

#include "meta_decl.H"
#include "meta_defn.H"

#endif /* !META_H__ */
