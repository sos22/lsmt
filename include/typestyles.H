#ifndef TYPESTYLES_H__
#define TYPESTYLES_H__

#include <string>

#include "typestyles_decl.H"

// Types with a typestyle member are easy. Pretty much means meta<>
// types.
template <typename t> typename t::typestyle gettypestyle(t ***);
typestyles::fundamental gettypestyle(char ***);
typestyles::fundamental gettypestyle(unsigned char ***);
typestyles::fundamental gettypestyle(short ***);
typestyles::fundamental gettypestyle(unsigned short ***);
typestyles::fundamental gettypestyle(int ***);
typestyles::fundamental gettypestyle(unsigned int ***);
typestyles::fundamental gettypestyle(long ***);
typestyles::fundamental gettypestyle(unsigned long ***);
typestyles::nonmeta gettypestyle(std::string ***);
typestyles::nonmeta gettypestyle(std::type_info ***);                                                                

#endif /* !TYPESTYLES_H__ */
