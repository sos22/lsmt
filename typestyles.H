#ifndef TYPESTYLES_H__
#define TYPESTYLES_H__

#include <string>

/* Style of type, for our meta framework. Either meta, for things
 * which inherit from meta<>, nonmeta, for classes which don't, or
 * fundamental for types like int or char which we can treat as a
 * fixed size block of bytes. Extract the style of type T with
 * gettypestyle(T***), which can't actually be called but the return
 * type is useful. */
namespace typestyles {
struct meta {};
struct nonmeta {};
struct fundamental {}; }
// Types with a typestyle member are easy. Pretty much means meta<>
// types.
template <typename t> typename t::typestyle gettypestyle(t ***);
typestyles::fundamental gettypestyle(char ***);
typestyles::fundamental gettypestyle(unsigned char ***);
typestyles::fundamental gettypestyle(short ***);
typestyles::fundamental gettypestyle(unsigned short ***);
typestyles::fundamental gettypestyle(int ***);
typestyles::fundamental gettypestyle(unsigned int ***);
typestyles::fundamental gettypestyle(long ***);
typestyles::fundamental gettypestyle(unsigned long ***);
typestyles::nonmeta gettypestyle(std::string ***);

#endif /* !TYPESTYLES_H__ */
