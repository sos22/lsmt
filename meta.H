#ifndef META_H__
#define META_H__

#include <assert.h>

#include "randomise.H"
#include "serialise.H"
#include "typestyles.H"


/* std::string mkjson(t const &) -- turn an arbitrary type into json,
 * as a string.  Works for meta and fundamental types. Can be made to
 * work for nonmetas if they implement nonmetatypes::mkjson. */
namespace nonmetatypes {
template <typename t> std::string mkjson(t const &); }
namespace fundamentaltypes { std::string mkjson(int); }
template <typename t> std::string mkjson(t const & what,
                                         typestyles::nonmeta) {
    return nonmetatypes::mkjson(what); }
template <typename t> std::string mkjson(t const & what,
                                         typestyles::fundamental) {
    return fundamentaltypes::mkjson(what); }
template <typename t> std::string mkjson(t const & what,
                                         typestyles::meta) {
    return what.json(); }
template <typename t> std::string mkjson(t const & what) {
    using sty = decltype(gettypestyle((t ***)nullptr));
    return mkjson(what, sty()); }

struct order {
    // Negative -> less than, 0 -> equal, positive -> greater than
private: int inner;
public:  explicit order(int _inner) : inner(_inner) { }
public:  const char * name() const {
    if (inner < 0) return "LT";
    else if (inner == 0) return "EQ";
    else return "GT"; }
public:  bool operator <(order o) const {
    if (o.inner < 0) return false;
    else if (inner < 0) return true;
    else if (o.inner == 0) return false;
    else if (inner == 0) return true;
    else return false; }
public: operator bool() const { return !!inner; } };

/* Type t can inherit from meta<t> to get some useful operations for free:
 *
 * -- serialise
 * -- deserialise
 * -- mkjson
 * -- randomise
 * -- operator <
 *
 * If you do, you need to implement a visit() method, which needs to give
 * the meta framework all of your member pointers. e.g.:
 *
 * struct foo : meta<foo> {
 *    int field1;
 *    std::string field2;
 *    some_other_type field3;
 *    template <typename visitor> static bool visit(visitor && v) {
 *       return v("field1", &foo::field1) &&
 *              v("field2", &foo::field2) &&
 *              v("field3", &foo::field3); } };
 *
 * The visitor should be called once for each field, giving it a
 * string description of the field (mostly useful for json) and a
 * pointer-to-member to the field. The visitor returns a bool, which
 * is either true to keep going or false to stop early. */
template <typename what>
struct meta {
    // Tag so that gettypestyle can easily recognise meta<> classes,
    // doing the SFINAE trick.
public: using typestyle = typestyles::meta;

    // Most things in this class are private, intended only to help
    // implement APIs on the matching interface class, so they're
    // private here and we're friends with the interfaces.
    friend struct deserialiser;
    friend struct randomiser;
    friend struct serialiser;

private: what const & inner() const { return *static_cast<what const *>(this); }
private: what & inner() { return *static_cast<what *>(this); }
private: void serialise(serialiser & t) const {
    what::visit([&](const char *, auto val) {
            t.serialise(inner().*val);
            return true; }); }
private: void deserialise(deserialiser & t) {
    what::visit(
        [&](const char *, auto val) {
            t.deserialise(inner().*val);
            return true; });}
private: void randomise(randomiser & t) {
    what::visit(
        [&](const char *, auto val) {
            t.randomise(inner().*val);
            return true; }); }

    // mkjson() is a function template, rather than a class, so friend
    // declaration is tricky. Just make it public instead.
public:  std::string json() const {
    std::string acc;
    acc.append("{");
    what::visit([&](const char * name, auto val) {
            acc.append("\"");
            acc.append(name);
            acc.append("\": ");
            auto v(mkjson(inner().*val));
            acc.append(v);
            acc.append(", ");
            return true; });
    if (acc.size() > 1) acc.erase(acc.begin() + acc.size() - 2, acc.end());
    acc.append("}");
    return acc; }

    // operators pretty much have to be public, or else what's the
    // point?
public:  bool operator<(const what & o) const {
    bool lt = false;
    what::visit(
        [&](const char *, auto val) {
            if (inner().*val < o.*val) {
                lt = true;
                return false; }
            else if (o.*val < inner().*val) {
                lt = false;
                return false; }
            else return true; });
    return lt; } };

// Assume we already have operator <. Derive some more relational
// operators.
template <typename a> bool operator>=(const a & x, const a & y) {
    return !(x < y); }
template <typename a> bool operator>(const a & x, const a & y) {
    return y < x; }
template <typename a> bool operator<=(const a & x, const a & y) {
    return !(y < x); }

template <typename a> bool operator==(const a & x, const a & y) {
    return !(x < y) && !(y < x); }
template <typename a> bool operator!=(const a & x, const a & y) {
    return !(x == y); }

#endif /* !META_H__ */
