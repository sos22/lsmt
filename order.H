#ifndef ORDER_H__
#define ORDER_H__

#include "typestyles.H"

struct order;

namespace nonmetatypes {
template <typename t> order compare(t const &, t const &); }

struct order {
    // -1 -> less than, 0 -> equal, +1 -> greater than
public:  int inner;
private: explicit constexpr order(int _inner) : inner(_inner) { }
public:  static const order lt;
public:  static const order eq;
public:  static const order gt;
public:  const char * name() const {
    if (inner < 0) return "LT";
    else if (inner == 0) return "EQ";
    else return "GT"; }
public: operator bool() const { return !!inner; } };
typestyles::nonmeta gettypestyle(order ***);

// Simple wrapper for compare operators. Has no state, so pretty
// pointless on the face of it, but makes it easier to get the right
// overload resolution and public/private.
struct comparer {
public: template <typename t> static order compare(const t & a, const t & b) {
    using sty = decltype(gettypestyle((t ***)nullptr));
    return compare(a, b, sty()); }
private: template <typename t> static order compare(const t & a,
                                                    const t & b,
                                                    typestyles::meta) {
    return a.compare(b); }
private: template <typename t> static order compare(t a,
                                                    t b,
                                                    typestyles::fundamental){
    if (a < b) return order::lt;
    else if (a == b) return order::eq;
    else return order::gt; }
private: template <typename t> static order compare(const t & a,
                                                    const t & b,
                                                    typestyles::nonmeta) {
    return nonmetatypes::compare(a, b); } };

// Assume we already have comparer::compare. Derive ordering
// operators.
template <typename a> bool operator< (const a & x, const a & y) {
    return comparer::compare(x, y) <  order::eq; }
template <typename a> bool operator<=(const a & x, const a & y) {
    return comparer::compare(x, y) <= order::eq; }
template <typename a> bool operator==(const a & x, const a & y) {
    return comparer::compare(x, y) == order::eq; }
template <typename a> bool operator>=(const a & x, const a & y) {
    return comparer::compare(x, y) >= order::eq; }
template <typename a> bool operator> (const a & x, const a & y) {
    return comparer::compare(x, y) >  order::eq; }

// Special case for unorderable types with operator== but no full compare.
template <typename a> bool operator!=(const a & x, const a & y) {
    return !(x == y); }

// Can't use template-generate operators for order itself, so do it by
// hand.
static inline bool operator< (const order & x, const order & y) {
    return x.inner <  y.inner; }
static inline bool operator<=(const order & x, const order & y) {
    return x.inner <= y.inner; }
static inline bool operator==(const order & x, const order & y) {
    return x.inner == y.inner; }
static inline bool operator>=(const order & x, const order & y) {
    return x.inner >= y.inner; }
static inline bool operator> (const order & x, const order & y) {
    return x.inner >  y.inner; }

#endif /* !ORDER_H__ */
