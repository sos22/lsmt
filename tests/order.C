#include "meta.H"

#include <iostream>

using namespace std;

std::ostream & operator<<(std::ostream & a, order b) { return a << b.name(); }

int main() {
#define opers(iter)                             \
    iter(<)                                     \
    iter(<=)                                \
    iter(==)                                \
    iter(!=)                                \
    iter(>=)                                \
    iter(>)

    cout << "  \t  \t";
#define one(name) cout << #name << "\t";
    opers(one)
#undef one
    cout << "bool\n";
    for (auto a : {order(-1), order(0), order(1)}) {
        for (auto b : {order(-1), order(0), order(1)}) {
            cout << a << "\t" << b << "\t";
#define one(name) cout << (a name b) << "\t";
            opers(one)
#undef one
            if (b == order(-1)) cout << (bool)(a);
            cout << "\n";
        }
    }
    assert((order(0) ?: order(1)) == order(1));
    assert((order(-1) ?: order(1)) == order(-1));
    return 0; }
