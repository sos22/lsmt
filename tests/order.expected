  	  	<	<=	==	!=	>=	>	bool
LT	LT	0	1	1	0	1	0	1
LT	EQ	1	1	0	1	0	0	
LT	GT	1	1	0	1	0	0	
EQ	LT	0	0	0	1	1	1	0
EQ	EQ	0	1	1	0	1	0	
EQ	GT	1	1	0	1	0	0	
GT	LT	0	0	0	1	1	1	1
GT	EQ	0	0	0	1	1	1	
GT	GT	0	1	1	0	1	0	
